### DISCUSSION TITLE
24.04 HW Acceleration with mpv

### DISCUSSION THREAD

**Discussion by MyVizDrake** [Joshua-Riek/ubuntu-rockchip#931] [Q&A]
Running 24.04 on a Rock5A ...

I had an MKV file that seemed to be SW decoding only (very high utilization on all 8x cores.  I converted it to mp4 H.264(x264) and still seeing the same high CPU utilization.

I suspect the issue is my video file is in a format that is not supported for HW acceleration.

```
here is the output from CLI for the MP4
<user>@rock5a-s:~/Videos$ mpv video.mp4 
 (+) Video --vid=1 (*) (h264 720x368 23.976fps)
 (+) Audio --aid=1 --alang=eng (*) (aac 2ch 48000Hz)
libEGL warning: MESA-LOADER: failed to open rknpu: /usr/lib/dri/rknpu_dri.so: cannot open shared object file: No such file or directory (search paths /usr/lib/aarch64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)

[vo/gpu/opengl] Suspected software renderer or indirect context.
Cannot load libcuda.so.1
[ffmpeg] AVHWDeviceContext: Cannot load libcuda.so.1
[ffmpeg] AVHWDeviceContext: Could not dynamically load CUDA
Failed to open VDPAU backend libvdpau_nvidia.so: cannot open shared object file: No such file or directory
Using hardware decoding (rkmpp).
rga_api version 1.10.0_[8]
[autoconvert] HW-downloading from drm_prime
AO: [pipewire] 48000Hz stereo 2ch floatp
[autoconvert] HW-downloading from drm_prime
VO: [gpu] 720x368 => 853x368 nv12
AV: 00:00:05 / 01:37:51 (0%) A-V:  0.000

Exiting... (Quit)
```

here is the output from CLI for the MKV
```
<user>@rock5a-s:~/Videos$ mpv video.mkv 
 (+) Video --vid=1 (mpeg2video 720x480 29.970fps)
 (+) Audio --aid=1 --alang=eng (*) 'Surround 5.1' (ac3 6ch 48000Hz)
     Audio --aid=2 --alang=fre 'Surround 5.1' (ac3 6ch 48000Hz)
     Audio --aid=3 --alang=spa 'Surround 5.1' (ac3 6ch 48000Hz)
     Audio --aid=4 --alang=eng 'Stereo' (ac3 2ch 48000Hz)
     Audio --aid=5 --alang=eng 'Surround 5.1' (ac3 6ch 48000Hz)
     Audio --aid=6 --alang=eng 'Stereo' (ac3 2ch 48000Hz)
 (*) Subs  --sid=1 --slang=eng (*) [F] (dvd_subtitle)
     Subs  --sid=2 --slang=fre (dvd_subtitle)
     Subs  --sid=3 --slang=spa (dvd_subtitle)
     Subs  --sid=4 --slang=eng (dvd_subtitle)
     Subs  --sid=5 --slang=fre (dvd_subtitle)
     Subs  --sid=6 --slang=spa (dvd_subtitle)
     Subs  --sid=7 --slang=eng (subrip)
libEGL warning: MESA-LOADER: failed to open rknpu: /usr/lib/dri/rknpu_dri.so: cannot open shared object file: No such file or directory (search paths /usr/lib/aarch64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri, suffix _dri)

[vo/gpu/opengl] Suspected software renderer or indirect context.
[ffmpeg] Impossible to convert between the formats supported by the filter 'mpv_src_default_in' and the filter 'auto_scale_0'
[lavfi] failed to configure the filter graph
Disabling filter scale_rkrga.00 because it has failed.
AO: [pipewire] 48000Hz 5.1(side) 6ch floatp
VO: [gpu] 720x480 => 853x480 yuv420p
AV: 00:00:06 / 01:37:51 (0%) A-V:  0.000 Dropped: 8
[osd/libass] fontselect: Using default font family: (mpv-osd-symbols, 400, 0) -> /usr/share/fonts/truetype/noto/NotoSans-Regular.ttf, 0, NotoSans-Regular
AV: 00:00:17 / 01:37:51 (0%) A-V:  0.000 Dropped: 511

Exiting... (Quit)
```

htop

![image](https://github.com/user-attachments/assets/8b2bc37b-39fb-4f8f-94a6-8f8e7697159f)

What should I use to encode the videos or the mpv command to enable HW accelerated video playback on this SBC?  I thought MPV was setup for HW Acceleration by default but am now not sure.

**Comment by Joshua-Riek** [comment]
It's highly likely that you installed something or made a system modification that broke hardware acceleration. I'm working on updating ffmpeg, mpp, and mpv today and have not found any issues. 

What kernel version are you running and can you post the video file?

**Reply by MyVizDrake** [reply]
I fear the "enable the Radxa Display 10FHD" thing that broke Chromium HW Acceleration(#861) is at work here as well ...  at the time you had indicated that on your 5B it only looked like it impacted Chromium ....

Kernel is Linux rock5a-s 6.1.0-1020-rockchip

Again, not discrediting user error. Could not get MKV to play with MPV from the applications menu so started via CLI and noticed the errors and though I would ask here 1st before posting an Issue. Appreciate you taking the time to look into this!

**Reply by Joshua-Riek** [reply]
Now that is interesting, this could point to a problem with mpp, when i'm done upgrading ffmpeg, mpp, and mpv ill see what I can find.

**Reply by nyanmisaka** [reply]
MPV requires the EGL interface provided by Mesa to import FFmpeg decoded images from the VPU to the GPU via FD(dma-buf) for rendering and display. Software rendering relies on the CPU and cannot do this efficiently...
```
[vo/gpu/opengl] Suspected software renderer or indirect context.
```

So enabling Radxa Display Panel causing Mesa to fall back to software rendering is unlikely to be caused by MPP. It is more like a compatibility issue in the panfork or vop driver.
```
Graphics Feature Status
Canvas: Software only, hardware acceleration unavailable
Canvas out-of-process rasterization: Disabled
Direct Rendering Display Compositor: Disabled
Compositing: Software only. Hardware acceleration disabled
Multiple Raster Threads: Enabled
OpenGL: Unavailable
Rasterization: Unavailable
Raw Draw: Disabled
Video Decode: Unavailable
Video Encode: Software only. Hardware acceleration disabled
Vulkan: Disabled
WebGL: Unavailable
WebGL2: Unavailable
WebGPU: Disabled
```

**Reply by Joshua-Riek** [reply]
Interesting, I do think its caused by Panfork since the GPU works with X11 and the Radx Display Panel but not Wayland.
